### ISSUE TITLE
Benchmark hightlighting struct with 1000 fields regressed. 

### ISSUE THREAD

**Issue by matklad** [rust-analyzer/rust-analyzer#8377]
https://rust-lang.zulipchat.com/#narrow/stream/185405-t-compiler.2Frust-analyzer/topic/Field.20resolution.20in.20hir_ty/near/233349001

**Comment by jonas-schievink** [comment]
Good commit: https://github.com/rust-analyzer/rust-analyzer/commit/f7156cb0aeaba8fe32c381a2d676b35d2c86f46f

Bad commit: https://github.com/rust-analyzer/rust-analyzer/commit/74711deb724e0a6c84a221bfdfa15eaedbdb2300

**Comment by matklad** [comment]
Better command line:

```
env RUN_SLOW_TESTS=1 cargo test -q --package ide --lib -- syntax_highlighting::tests::benchmark_syntax_highlighting_long_struct --exact --nocapture
```

**Comment by eminence** [comment]
Bisect says 636de3c709a7c86a1d3a870dc5dc3566310e9d92 is the first bad commit (from #8112)

**Comment by jonas-schievink** [comment]
That commit takes much less time than the more recent 74711de, so I think there might be more to it.

#8386 should fix most of the regression introduced by 636de3c.

**Comment by jonas-schievink** [comment]
> That commit takes much less time than the more recent [74711de](https://github.com/rust-analyzer/rust-analyzer/commit/74711deb724e0a6c84a221bfdfa15eaedbdb2300), so I think there might be more to it.

Hmm, I think this might have been a fluke.
